presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -141,43 +144,70 @@ private PlanNodeStatsEstimate computeRightJoinStats(
+            if (!node.getFilter().isPresent()) {
+                return crossJoinStats;
+            }
+            return filterStatsCalculator.filterStats(crossJoinStats, node.getFilter().get(), session, types);
+        }
+
+        PlanNodeStatsEstimate equiJoinEstimate = filterByEquiJoinClauses(crossJoinStats, node.getCriteria(), session, types);
+
+        if (equiJoinEstimate.isOutputRowCountUnknown()) {
+            return PlanNodeStatsEstimate.unknown();
+        }
+
+        if (!node.getFilter().isPresent()) {
+            return equiJoinEstimate;
+        }
+
+        PlanNodeStatsEstimate filteredEquiJoinEstimate = filterStatsCalculator.filterStats(equiJoinEstimate, node.getFilter().get(), session, types);
+
+        if (filteredEquiJoinEstimate.isOutputRowCountUnknown()) {
+            return equiJoinEstimate.mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+        }
+
+        return filteredEquiJoinEstimate;
+    }
+
+    private PlanNodeStatsEstimate filterByEquiJoinClauses(
+            PlanNodeStatsEstimate stats,
+            Collection<EquiJoinClause> clauses,
+            Session session,
+            TypeProvider types)
+    {
+        checkArgument(!clauses.isEmpty(), "clauses is empty");
+        PlanNodeStatsEstimate result = PlanNodeStatsEstimate.unknown();
         // Join equality clauses are usually correlated. Therefore we shouldn't treat each join equality
         // clause separately because stats estimates would be way off. Instead we choose so called
         // "driving clause" which mostly reduces join output rows cardinality and apply UNKNOWN_FILTER_COEFFICIENT
         // for other (auxiliary) clauses.
-        PlanNodeStatsEstimate equiJoinClausesFilteredStats = IntStream.range(0, equiJoinClauses.size())
-                .mapToObj(drivingClauseId -> {
-                    EquiJoinClause drivingClause = equiJoinClauses.get(drivingClauseId);
-                    List<EquiJoinClause> remainingClauses = copyWithout(equiJoinClauses, drivingClauseId);
-                    return filterByEquiJoinClauses(crossJoinStats, drivingClause, remainingClauses, session, types);
-                })
-                .min(comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
-                .orElse(crossJoinStats);
-
-        return node.getFilter()
-                .map(filter -> filterStatsCalculator.filterStats(equiJoinClausesFilteredStats, filter, session, types))
-                .orElse(equiJoinClausesFilteredStats);
-    }
+        Queue<EquiJoinClause> remainingClauses = new LinkedList<>(clauses);
+        EquiJoinClause drivingClause = remainingClauses.poll();
+        for (int i = 0; i < clauses.size(); i++) {
+            PlanNodeStatsEstimate estimate = filterByEquiJoinClauses(stats, drivingClause, remainingClauses, session, types);
+            if (result.isOutputRowCountUnknown() || (!estimate.isOutputRowCountUnknown() && estimate.getOutputRowCount() < result.getOutputRowCount())) {
```

Comment:
I don't see how this implements that:
https://github.com/prestodb/presto/pull/11625#issuecomment-426688787
```
For the equi joins at least one equi criteria must be "reasonably" estimated.
For non equi joins at least one filter conjuct must be "reasonably" estimated.
```